vcell-client/src/main/java/cbit/vcell/mapping/gui/MolecularTypeSpecsTableModel.java

```diff
@@ -163,45 +190,119 @@ public Object getValueAt(int row, int col) {
+					return;
+				}
+				Coordinate c = sas.getCoordinate();
+				if(c.getX() != res) {
+					c = new Coordinate(res, c.getY(), c.getZ());
+					sas.setCoordinate(c);
+					scs.firePropertyChange(SpeciesContextSpec.PROPERTY_NAME_SITE_ATTRIBUTE, null, sas);
+				}
+			}
+			return;
+		case COLUMN_Y:
+			if (aValue instanceof String) {
+				if(sas == null) {
+					// TODO: is this necessary?
+					sas = new SiteAttributesSpec(scs, mcp, scs.getSpeciesContext().getStructure());
+				}
+				String newExpressionString = (String)aValue;
+				double res = 0.0;
+				try {
+					res = Double.parseDouble(newExpressionString);
+				} catch(NumberFormatException e) {
+					return;
+				}
+				Coordinate c = sas.getCoordinate();
+				if(c.getX() != res) {
+					c = new Coordinate(c.getX(), res, c.getZ());
+					sas.setCoordinate(c);
+					scs.firePropertyChange(SpeciesContextSpec.PROPERTY_NAME_SITE_ATTRIBUTE, null, sas);
+				}
+			}
+			return;
+		case COLUMN_Z:
+			if (aValue instanceof String) {
+				if(sas == null) {
+					sas = new SiteAttributesSpec(scs, mcp, scs.getSpeciesContext().getStructure());
+				}
+				String newExpressionString = (String)aValue;
+				double res = 0.0;
+				try {
+					res = Double.parseDouble(newExpressionString);
+				} catch(NumberFormatException e) {
+					return;
+				}
+				Coordinate c = sas.getCoordinate();
+				if(c.getX() != res) {
+					c = new Coordinate(c.getX(), c.getY(), res);
+					sas.setCoordinate(c);
+					// updates the Length in the links table (LinkSpecsTableModel)
+					scs.firePropertyChange(SpeciesContextSpec.PROPERTY_NAME_SITE_ATTRIBUTE, null, sas);
+				}
 			}
 			return;
 		case COLUMN_RADIUS:
 			if (aValue instanceof String) {
 				String newExpressionString = (String)aValue;
 				double result = Double.parseDouble(newExpressionString);
-				SiteAttributesSpec sas = getSpeciesContextSpec().getSiteAttributesMap().get(mcp);
 				if(sas == null) {
-					sas = new SiteAttributesSpec(fieldSpeciesContextSpec, mcp, getSpeciesContextSpec().getSpeciesContext().getStructure());
+					sas = new SiteAttributesSpec(scs, mcp, scs.getSpeciesContext().getStructure());
 				}
 				sas.setRadius(result);
+				scs.firePropertyChange(SpeciesContextSpec.PROPERTY_NAME_SITE_ATTRIBUTE, null, sas);
 			}
 			return;
 		case COLUMN_DIFFUSION:
 			if (aValue instanceof String) {
 				String newExpressionString = (String)aValue;
 				double result = Double.parseDouble(newExpressionString);
-				SiteAttributesSpec sas = getSpeciesContextSpec().getSiteAttributesMap().get(mcp);
 				if(sas == null) {
-					sas = new SiteAttributesSpec(fieldSpeciesContextSpec, mcp, getSpeciesContextSpec().getSpeciesContext().getStructure());
+					sas = new SiteAttributesSpec(scs, mcp, scs.getSpeciesContext().getStructure());
 				}
 				sas.setDiffusionRate(result);
+				scs.firePropertyChange(SpeciesContextSpec.PROPERTY_NAME_SITE_ATTRIBUTE, null, sas);
+				return;
+			}
+		case COLUMN_COLOR:
+			if (aValue instanceof NamedColor) {
```

Comment:
there is now a convenience for downcasting in java, you should probably start using it (more convenient and less verbose).

replace:

if (aValue instanceof NamedColor) {
     NamedColor namedColor = (NamedColor)aValue;
				
with the equivalent statement:

if (aValue instanceof NamedColor namedColor) {

Reply:
I know about it, you showed me, but old habits die hard... I will try to remember to use it